runtime/src/accounts_db.rs

```diff
@@ -782,40 +779,47 @@ impl AccountsDB {
         }
     }
 
-    fn handle_reclaims_maybe_cleanup(&self, reclaims: SlotSlice<AccountInfo>) {
-        let mut dead_accounts = Measure::start("reclaims::remove_dead_accounts");
-        let dead_slots = self.remove_dead_accounts(reclaims);
-        dead_accounts.stop();
-        let dead_slots_len = {
-            let mut dead_slots_w = self.dead_slots.write().unwrap();
-            dead_slots_w.extend(dead_slots);
-            dead_slots_w.len()
-        };
-        if dead_slots_len > 5000 {
-            self.process_dead_slots(None);
+    // Removes the accounts in the input `reclaims` from the tracked "count" of
+    // their corresponding  storage entries. Note this does not actually free
+    // the memory from the storage entries until all the storage entries for
+    // a given slot `S` are empty, at which point `process_dead_slots` will
+    // remove all the storage entries for `S`.
+    //
+    /// # Arguments
+    /// * `reclaims` - The accounts to remove from storage entries' "count"
+    /// * `expected_slot` - A correctness assertion. If this is equal to `Some(S)`,
+    /// then the function will check that the only slot being cleaned up in `reclaims`
+    /// is the slot == `S`. This is true for instance when `handle_reclaims` is called
+    /// from store or slot shrinking, as those should only touch the slot they are
+    /// currently storing to or shrinking.
+    /// * `is_cleanup_possible` - A correctness assertion. If this is equal to
+    /// `false`, the function will check that no slots are cleaned up/removed via
+    /// `process_dead_slots`. For instance, on store, no slots should be cleaned up,
+    /// but during the background clean accounts purges accounts from old rooted slots,
+    /// so outdated slots may be removed.
+    fn handle_reclaims(
+        &self,
+        reclaims: SlotSlice<AccountInfo>,
+        expected_slot: Option<Slot>,
+        is_cleanup_possible: bool,
+    ) {
+        if !reclaims.is_empty() {
+            let dead_slots = self.remove_dead_accounts(reclaims, expected_slot);
+            if !is_cleanup_possible {
+                assert!(dead_slots.is_empty());
+            } else if let Some(expected_slot) = expected_slot {
+                assert!(dead_slots.len() <= 1);
+                if dead_slots.len() == 1 {
+                    assert!(dead_slots.contains(&expected_slot));
+                }
+            }
+            self.process_dead_slots(&dead_slots);
```

Comment:
Let's short-circuit for the case of `dead_slots.empty()` here for readability and potential further throughput for the `store` pipeline: 

```rust
if !dead_slots.is_empty() {
   self.process_dead_slots(&dead_slots);
}
```

Then we can remove the eye-soar big `if !dead_slots.is_empty()` in `clean_dead_slots()`.
Also, we will skip a read lock for `accounts_index` and a WRITE lock for  `storage` in `purge_slots()` (which currently doesn't have the large emptiness check). (Or we already could skip if rustc is smart enough.)

Last bonus nits: also `purge_slots` and `clean_dead_slots` methods don't need to be `pub`)

Reply:
good point, done!